File: [23.12]Design-Pattern/Java/src/behavioral/observer/Store.java
package behavioral.observer;

import java.util.ArrayList;
import java.util.List;

public class Store implements Subject {

    private List<Observer> subscirbers = new ArrayList<>();

    @Override
    public void registerObserver(Observer observer) {
        subscirbers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        subscirbers.remove(observer);
    }

    @Override
    public void sendNotice(String msg) {
        System.out.println("[구독자 메시지 전달 시작]");
        for (Observer o :
                subscirbers) {
            o.receiveNotice(msg);
        }
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/behavioral/obsever/Store.kt
package behavioral.obsever

class Store: Subject {
    private val subscribers = mutableListOf<Observer>()

    override fun registerObserver(observer: Observer) {
        subscribers.add(observer)
    }

    override fun removeObserver(observer: Observer) {
        subscribers.remove(observer)
    }

    override fun sendNotice(msg: String) {
        println("[구독자 송신 목록]")
        subscribers.forEach{
            it.receiveNotice(msg)
        }
    }
}